### PR TITLE
[8.0] [IMP] load_needaction response time

### PR DESCRIPTION
**Pull request by mvaled**
The first `/web/menu/load_needaction` AJAX request contains every menu in the app, even those that are not shown upfront.  This may take several too much time to calculate.  Furthermore, since another `load_needaction` is performed when you switch to another main menu, fetching that data upfront is wasteful. 

The following pictures show, first, a local Odoo instance running without the patches and then the same instance with the patch to `chrome.js` applied.  Though hardly a benchmark, they show there's much room for improvement. 

Before applying the chrome.js patch

After the patch is applied

Notice that both requests pass the bulk of their time waiting for the response first bytes to arrive.

I haven't found any interference so far, but I would like more feedback about the "menu_bound" event.

### CODE REVIEW

File: openerp/addons/base/ir/ir_ui_menu.py [hunk 1]
```diff
@@ -329,20 +329,23 @@ def get_needaction_data(self, cr, uid, ids, context=None):
...
-                        res[menu.id]['needaction_enabled'] = obj._needaction
-                        res[menu.id]['needaction_counter'] = obj._needaction_count(cr, uid, dom, context=context)
+                obj = self.pool.get(menu.action.res_model, None)
+                if obj and obj._needaction:
+                    if menu.action.type == 'ir.actions.act_window':
+                        dom = menu.action.domain and eval(menu.action.domain, {'uid': uid}) or []
+                    else:
+                        dom = eval(menu.action.params_store or '{}', {'uid': uid}).get('domain')
+                    res[menu.id]['needaction_enabled'] = enabled = True
+                    res[menu.id]['needaction_counter'] = count = obj._needaction_count(cr, uid, dom, context=context)
```
Comment (xmo-odoo) [inline openerp/addons/base/ir/ir_ui_menu.py]: Why not just set `res[menu.id]` here (with a single literal dict)? That way no need for the temporary variables or for deleting it post-facto.

File: openerp/addons/base/ir/ir_ui_menu.py [hunk 2]
```diff
@@ -329,20 +329,23 @@ def get_needaction_data(self, cr, uid, ids, context=None):
...
-                if menu.action.res_model in self.pool:
-                    obj = self.pool[menu.action.res_model]
-                    if obj._needaction:
-                        if menu.action.type == 'ir.actions.act_window':
-                            dom = menu.action.domain and eval(menu.action.domain, {'uid': uid}) or []
-                        else:
-                            dom = eval(menu.action.params_store or '{}', {'uid': uid}).get('domain')
-                        res[menu.id]['needaction_enabled'] = obj._needaction
-                        res[menu.id]['needaction_counter'] = obj._needaction_count(cr, uid, dom, context=context)
+                obj = self.pool.get(menu.action.res_model, None)
```
Comment (xmo-odoo) [inline openerp/addons/base/ir/ir_ui_menu.py]: `None` is unnecessary, and is there *really* a chance of having an action of a current menu using a model which does not exist?

File: openerp/addons/base/ir/ir_ui_menu.py (hunk 1 above)
Comment (mvaled) [inline openerp/addons/base/ir/ir_ui_menu.py]: What I did here was to simply remove the double checking of self.pool: once for the containment and then for the actual getting.  I didn't want to mess too much with the previous code, for the patch were simpler to understand.

**Comment by mvaled** [comment]
Hi @tde-banana-odoo,

After a couple of days of this patch being deployed in our server.  I think the JS needs a bit of revision.

**Comment by mvaled** [comment]
Opps I closed this unintentionally:  I want to fix the JS stuff.

**Comment by mvaled** [comment]
Hi @tde-banana-odoo,

I've just added a new commit that addresses the comments before earlier by @xmo-odoo.  I still need to review the JS, though.

Best regards,
Manuel.

**Pull request closed without merging**